[Java] COMPATIBLE mode ClassCastException for unregistered Collection subclasses carrying extra fields (e.g. Hibernate PersistentSet)

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues describing this exact failure (related: #3621, but that covers a different symptom and its fix doesn't cover this case — see below).

### Version

1.2.0, 1.3.0 (not reproducible on 1.1.0)

### Component(s)

Java

### Minimal reproduce step

```java
import org.apache.fory.Fory;
import org.apache.fory.ThreadSafeFory;
import org.apache.fory.config.Language;
import org.hibernate.collection.spi.PersistentSet;

import java.util.HashSet;
import java.util.Set;

public class Repro {
  public static class Holder {
    Set<String> tags;
  }

  public static void main(String[] args) {
    ThreadSafeFory fory = Fory.builder()
        .withLanguage(Language.JAVA)
        .withRefTracking(true)
        .requireClassRegistration(false)
        .buildThreadSafeFory();

    HashSet<String> backing = new HashSet<>();
    backing.add("a");
    backing.add("b");

    Holder holder = new Holder();
    // any Set/Collection impl that is (a) unregistered, (b) not a java.util builtin, and
    // (c) has instance fields beyond its elements should reproduce this
    holder.tags = new PersistentSet<>(null, backing);

    byte[] bytes = fory.serialize(holder);
    fory.deserialize(bytes); // throws
  }
}
```

`org.hibernate.collection.spi.PersistentSet` is from `org.hibernate.orm:hibernate-core` (used here as a real-world example of a `Set` implementation carrying extra instance state alongside its elements — session, role, key, dirty flags, etc.).

### What did you expect to see?

Either:

1. The unregistered `Set` implementation round-trips (as it does on fory-core 1.1.0), or
2. A clear, actionable error at serialize or deserialize time, rather than an internal `ClassCastException` surfacing from generated code.

### What did you see instead?

```
org.apache.fory.exception.DeserializationException: Deserialize failed, read objects are: [Holder, null]
Caused by: java.lang.ClassCastException: class org.hibernate.collection.spi.PersistentSetForyRefCodecCompatible1_0 cannot be cast to class org.apache.fory.serializer.collection.CollectionLikeSerializer (org.hibernate.collection.spi.PersistentSetForyRefCodecCompatible1_0 is in unnamed module of loader org.apache.fory.util.ClassLoaderUtils$ByteArrayClassLoader; org.apache.fory.serializer.collection.CollectionLikeSerializer is in unnamed module of loader 'app')
	at org.hibernate.collection.spi.PersistentSetForyRefCodecCompatible1_0.readCollection$(PersistentSetForyRefCodecCompatible1_0.java)
	at org.hibernate.collection.spi.PersistentSetForyRefCodecCompatible1_0.readFields3$(PersistentSetForyRefCodecCompatible1_0.java)
	at org.hibernate.collection.spi.PersistentSetForyRefCodecCompatible1_0.read(PersistentSetForyRefCodecCompatible1_0.java)
	at org.apache.fory.context.ReadContext.readDataInternal(ReadContext.java)
	at org.apache.fory.context.ReadContext.readNonRef(ReadContext.java)
	at org.apache.fory.context.ReadContext.readRef(ReadContext.java)
	at org.apache.fory.Fory.deserialize(Fory.java)
```

The same generated-code pattern appears at the enclosing bean's own field-read site, not just inside `PersistentSet`'s codec:

```java
// generated for Holder.tags (declared java.util.Set<String>)
private Object readCollection(ReadContext readContext, MemoryBuffer memoryBuffer) {
    CollectionLikeSerializer collectionSerializer =
        (CollectionLikeSerializer) _f_typeResolver.readTypeInfo(readContext, setTypeInfoHolder).getSerializer();
    ...
}
```

### Root cause (as far as I could determine)

The generated field-read code for a `Set`/`Collection`-typed field unconditionally casts whatever serializer resolves for the field's concrete runtime type to `CollectionLikeSerializer`. Because `PersistentSet` is unregistered and has instance fields beyond its Set elements, its own serializer is built via `CompatibleCodecBuilder` and extends `GeneratedCompatibleSerializer` — a plain bean codec that does **not** implement `CollectionLikeSerializer`. The cast above then fails.

Bisected by pinning only the `fory-core` dependency version against the same reproduction:

- `1.1.0` — passes (a single non-"Compatible" codec is generated and shared between read and write for the concrete class, so no such cast happens against a bean codec)
- `1.2.0` — fails with the exact exception above
- `1.3.0` — fails with the exact exception above

This lines up with the 1.2.0 release note "default compatible mode for native serialization" as the likely point where read-time codec selection changed for fields like this.

### Anything else?

This looks like the same underlying gap as #3621 (COMPATIBLE mode mishandling unregistered, non-`java.util` Collection/Map implementations), surfacing with a different symptom — a hard `ClassCastException` here instead of silent data loss there. The fix that closed #3621 (#3628) only special-cased containers implementing `Externalizable`; `PersistentSet` (and presumably other custom collection wrappers carrying their own extra state) is not `Externalizable`, so it still falls into the same general gap.

--- a/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
+++ b/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
@@ -2017,7 +2017,9 @@ public class ClassResolver extends TypeResolver {
     TypeDef typeDef;
     Preconditions.checkArgument(
         serializerClass != UnknownClassSerializers.UnknownStructSerializer.class);
-    if (needToWriteTypeDef(serializerClass)) {
+    if (needToWriteTypeDef(serializerClass) || needsCollectionFieldTypeDef(serializerClass)) {
+      // Default collection/map serializers remain non-struct roots, but their wrapper fields still
+      // need TypeDef metadata so remote compatible readers can evolve those fields.
       typeDef = typeDefMap.computeIfAbsent(typeInfo.type, cls -> TypeDef.buildTypeDef(this, cls));
     } else {
       // Some type will use other serializers such MapSerializer and so on.
@@ -2030,6 +2032,11 @@ public class ClassResolver extends TypeResolver {
     }
     typeInfo.typeDef = typeDef;
     return typeDef;
+  }
+
+  private static boolean needsCollectionFieldTypeDef(Class<? extends Serializer> serializerClass) {
+    return serializerClass == CollectionSerializers.DefaultJavaCollectionSerializer.class
+        || serializerClass == MapSerializers.DefaultJavaMapSerializer.class;
   }
 
   /**

--- a/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
+++ b/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
@@ -954,7 +954,13 @@ public class ClassResolver extends TypeResolver {
     if (hasFieldMetadata) {
       // Preserve the normal TypeInfo/name cache so locally generated or dynamically registered
       // classes can be resolved when the TypeDef is decoded by the same resolver.
-      getTypeIdForTypeDef(cls);
+      int typeId = getTypeIdForTypeDef(cls);
+      if (usesNonStructTypeDef(cls)) {
+        // Field metadata does not make collection/map/custom-serializer classes struct-owned.
+        // Their natural serializers may delegate field IO internally, but the TypeDef root kind
+        // must still resolve to the non-struct serializer family on the reader.
+        return normalizeTypeDefRootTypeId(cls, typeId);
+      }
       return getFieldMetadataTypeIdForTypeDef(cls);
     }
     TypeInfo typeInfo = classInfoMap.get(cls);

--- a/java/fory-core/src/main/java/org/apache/fory/resolver/TypeResolver.java
+++ b/java/fory-core/src/main/java/org/apache/fory/resolver/TypeResolver.java
@@ -93,6 +93,8 @@ import org.apache.fory.serializer.UnknownClass;
 import org.apache.fory.serializer.UnknownClass.UnknownEmptyStruct;
 import org.apache.fory.serializer.UnknownClass.UnknownStruct;
 import org.apache.fory.serializer.UnknownClassSerializers;
+import org.apache.fory.serializer.collection.CollectionSerializers;
+import org.apache.fory.serializer.collection.MapSerializers;
 import org.apache.fory.type.Descriptor;
 import org.apache.fory.type.DescriptorBuilder;
 import org.apache.fory.type.DescriptorGrouper;
@@ -1136,7 +1138,10 @@ public abstract class TypeResolver {
     TypeInfo typeInfo;
     if (!typeDef.isStructSchemaKind()
         && !UnknownClass.class.isAssignableFrom(TypeUtils.getComponentIfArray(cls))) {
-      typeInfo = getTypeInfo(cls);
+      typeInfo =
+          typeDef.getFieldsInfo().isEmpty()
+              ? getTypeInfo(cls)
+              : getMetaSharedTypeInfo(typeDef, cls);
     } else if (ClassResolver.useReplaceResolveSerializer(cls)) {
       // For classes with writeReplace/readResolve, use their natural serializer
       // (ReplaceResolveSerializer) instead of CompatibleSerializer
@@ -1193,6 +1198,14 @@ public abstract class TypeResolver {
       if ((Types.isExtType(streamTypeId) && !Types.isExtType(localTypeInfo.typeId))
           || (Types.isUnionType(streamTypeId) && !Types.isUnionType(localTypeInfo.typeId))) {
         throw new SerializerUnregisteredException(typeDef.getClassName());
+      }
+      Serializer<?> serializer = newMetaSharedExtSerializer(cls, typeDef, localTypeInfo);
+      if (serializer != null) {
+        TypeInfo typeInfo =
+            new TypeInfo(this, cls, null, localTypeInfo.typeId, typeDef.getClassSpec().userTypeId);
+        typeInfo.typeDef = typeDef;
+        typeInfo.setSerializer(this, serializer);
+        return typeInfo;
       }
       return localTypeInfo;
     }
@@ -1299,6 +1312,22 @@ public abstract class TypeResolver {
       typeInfo.setSerializer(this, Serializers.newSerializer(this, cls, sc));
     }
     return typeInfo;
+  }
+
+  private Serializer<?> newMetaSharedExtSerializer(
+      Class<?> cls, TypeDef typeDef, TypeInfo localTypeInfo) {
+    if (typeDef.getFieldsInfo().isEmpty()) {
+      return null;
+    }
+    Serializer<?> serializer = localTypeInfo.getSerializer();
+    Class<?> serializerClass = serializer.getClass();
+    if (serializerClass == CollectionSerializers.DefaultJavaCollectionSerializer.class) {
+      return new CollectionSerializers.DefaultJavaCollectionSerializer<>(this, cls, typeDef);
+    }
+    if (serializerClass == MapSerializers.DefaultJavaMapSerializer.class) {
+      return new MapSerializers.DefaultJavaMapSerializer<>(this, cls, typeDef);
+    }
+    return null;
   }
 
   private Serializer<?> newGeneratedCompatibleSerializer(

--- a/java/fory-core/src/main/java/org/apache/fory/serializer/CompatibleLayerSerializerBase.java
+++ b/java/fory-core/src/main/java/org/apache/fory/serializer/CompatibleLayerSerializerBase.java
@@ -19,6 +19,7 @@
 
 package org.apache.fory.serializer;
 
+import org.apache.fory.annotation.Internal;
 import org.apache.fory.collection.IdentityObjectIntMap;
 import org.apache.fory.collection.ObjectIntMap;
 import org.apache.fory.context.CopyContext;
@@ -154,6 +155,18 @@ public abstract class CompatibleLayerSerializerBase<T> extends AbstractObjectSer
 
   public int getNumFields() {
     return allFields.length;
+  }
+
+  @Internal
+  public final TypeDef getLayerTypeDef() {
+    checkLayerSerializerMeta();
+    return layerTypeDef;
+  }
+
+  @Internal
+  public final Class<?> getLayerMarkerClass() {
+    checkLayerSerializerMeta();
+    return layerMarkerClass;
   }
 
   @SuppressWarnings("rawtypes")

--- a/java/fory-core/src/main/java/org/apache/fory/serializer/collection/ChildContainerSerializers.java
+++ b/java/fory-core/src/main/java/org/apache/fory/serializer/collection/ChildContainerSerializers.java
@@ -25,6 +25,7 @@ import java.lang.invoke.MethodHandle;
 import java.lang.reflect.Field;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
@@ -54,9 +55,11 @@ import org.apache.fory.memory.MemoryBuffer;
 import org.apache.fory.meta.TypeDef;
 import org.apache.fory.reflect.ReflectionUtils;
 import org.apache.fory.resolver.ClassResolver;
+import org.apache.fory.resolver.TypeInfo;
 import org.apache.fory.resolver.TypeResolver;
 import org.apache.fory.serializer.AbstractObjectSerializer;
 import org.apache.fory.serializer.CompatibleLayerSerializer;
+import org.apache.fory.serializer.CompatibleLayerSerializerBase;
 import org.apache.fory.serializer.FieldGroups;
 import org.apache.fory.serializer.FieldGroups.SerializationFieldInfo;
 import org.apache.fory.serializer.JavaSerializer;
@@ -645,12 +648,11 @@ public class ChildContainerSerializers {
     readAndCheckNumClassLayers(readContext, collection.getClass(), slotsSerializers.length);
     for (Serializer slotsSerializer : slotsSerializers) {
       if (slotsSerializer instanceof CompatibleLayerSerializer) {
-        CompatibleLayerSerializer compatibleSerializer =
-            (CompatibleLayerSerializer) slotsSerializer;
-        // Read layer class meta first if meta share is enabled
-        // This corresponds to writeLayerClassMeta() in CompatibleLayerSerializer.write()
+        CompatibleLayerSerializerBase compatibleSerializer =
+            (CompatibleLayerSerializerBase) slotsSerializer;
         if (typeResolver.getConfig().isMetaShareEnabled()) {
-          readAndSkipLayerClassMeta(readContext);
+          compatibleSerializer =
+              readLayerSerializer(readContext, typeResolver, compatibleSerializer);
         }
         compatibleSerializer.readAndSetFields(readContext, collection);
       } else {
@@ -686,16 +688,16 @@ public class ChildContainerSerializers {
   }
 
   /**
-   * Read and skip the layer class meta from buffer. This is used to skip over the class definition
-   * that was written by CompatibleLayerSerializer.writeLayerClassMeta(). For
-   * ChildContainerSerializers, we use the same serializer on both write and read sides, so we just
-   * need to skip the meta without actually parsing it.
+   * Read the layer class meta that was written by CompatibleLayerSerializer.writeLayerClassMeta().
    */
-  private static void readAndSkipLayerClassMeta(ReadContext readContext) {
+  private static CompatibleLayerSerializerBase readLayerSerializer(
+      ReadContext readContext,
+      TypeResolver typeResolver,
+      CompatibleLayerSerializerBase localSerializer) {
     MemoryBuffer buffer = readContext.getBuffer();
     MetaReadContext metaReadContext = readContext.getMetaReadContext();
     if (metaReadContext == null) {
-      return;
+      return localSerializer;
     }
     int indexMarker = buffer.readVarUInt32Small14();
     boolean isRef = (indexMarker & 1) == 1;
@@ -704,16 +706,47 @@ public class ChildContainerSerializers {
       if (index >= metaReadContext.readTypeInfos.size) {
         throw new ForyException("Invalid layer metadata reference id " + index);
       }
-      // Reference to previously read type - nothing more to read
-      return;
+      TypeInfo typeInfo = metaReadContext.readTypeInfos.get(index);
+      if (typeInfo == null) {
+        throw new ForyException("Invalid layer metadata reference id " + index);
+      }
+      return getLayerSerializer(typeResolver, localSerializer, typeInfo);
     }
-    // New type - need to read and skip the TypeDef bytes
     long id = buffer.readInt64();
-    TypeDef.skipTypeDef(buffer, id);
-    // Add a placeholder to keep readTypeInfos indices in sync with the write side's classMap.
-    // The write side (writeLayerClassMeta) adds layer marker classes to classMap which shares
-    // the same index space as writeSharedClassMeta. Without this placeholder, subsequent
-    // readSharedClassMeta reference lookups would use wrong indices.
-    metaReadContext.readTypeInfos.add(null);
+    TypeInfo typeInfo = readLayerTypeInfo(typeResolver, buffer, localSerializer, id);
+    metaReadContext.readTypeInfos.add(typeInfo);
+    return getLayerSerializer(typeResolver, localSerializer, typeInfo);
+  }
+
+  private static TypeInfo readLayerTypeInfo(
+      TypeResolver typeResolver,
+      MemoryBuffer buffer,
+      CompatibleLayerSerializerBase localSerializer,
+      long typeDefId) {
+    TypeDef localTypeDef = localSerializer.getLayerTypeDef();
+    byte[] encoded = TypeDef.readTypeDefBytes(typeResolver, buffer, typeDefId);
+    Class<?> layerClass = localSerializer.getType();
+    typeResolver.checkClassForDeserialization(layerClass);
+    TypeDef typeDef =
+        Arrays.equals(encoded, localTypeDef.getEncoded())
+            ? localTypeDef
+            : typeResolver.cacheRemoteTypeDef(TypeDef.readTypeDef(typeResolver, encoded));
+    return new TypeInfo(layerClass, typeDef);
+  }
+
+  private static CompatibleLayerSerializerBase getLayerSerializer(
+      TypeResolver typeResolver, CompatibleLayerSerializerBase localSerializer, TypeInfo typeInfo) {
+    Serializer<?> serializer = typeInfo.getSerializer();
+    if (serializer != null) {
+      return (CompatibleLayerSerializerBase) serializer;
+    }
+    CompatibleLayerSerializer<?> newSerializer =
+        new CompatibleLayerSerializer(
+            typeResolver,
+            localSerializer.getType(),
+            typeInfo.getTypeDef(),
+            localSerializer.getLayerMarkerClass());
+    typeInfo.setSerializer(newSerializer);
+    return newSerializer;
   }
 }

--- a/java/fory-core/src/main/java/org/apache/fory/serializer/collection/CollectionSerializers.java
+++ b/java/fory-core/src/main/java/org/apache/fory/serializer/collection/CollectionSerializers.java
@@ -56,12 +56,14 @@ import org.apache.fory.exception.DeserializationException;
 import org.apache.fory.exception.ForyException;
 import org.apache.fory.memory.MemoryBuffer;
 import org.apache.fory.memory.MemoryUtils;
+import org.apache.fory.meta.TypeDef;
 import org.apache.fory.reflect.FieldAccessor;
 import org.apache.fory.reflect.ReflectionUtils;
 import org.apache.fory.resolver.ClassResolver;
 import org.apache.fory.resolver.TypeInfo;
 import org.apache.fory.resolver.TypeInfoHolder;
 import org.apache.fory.resolver.TypeResolver;
+import org.apache.fory.serializer.CompatibleSerializer;
 import org.apache.fory.serializer.ExternalizableSerializer;
 import org.apache.fory.serializer.GraphMemoryEstimates;
 import org.apache.fory.serializer.ReplaceResolveSerializer;
@@ -1050,17 +1052,33 @@ public class CollectionSerializers {
     private Serializer<T> dataSerializer;
 
     public DefaultJavaCollectionSerializer(TypeResolver typeResolver, Class<T> cls) {
+      this(typeResolver, cls, null, true);
+    }
+
+    public DefaultJavaCollectionSerializer(
+        TypeResolver typeResolver, Class<T> cls, TypeDef typeDef) {
+      this(typeResolver, cls, typeDef, false);
+    }
+
+    private DefaultJavaCollectionSerializer(
+        TypeResolver typeResolver, Class<T> cls, TypeDef typeDef, boolean registerSerializer) {
       super(typeResolver, cls, false);
       Preconditions.checkArgument(
           !config.isXlang(),
           "Fory cross-language default collection serializer should use "
               + CollectionSerializer.class);
-      typeResolver.setSerializer(cls, this);
-      Class<? extends Serializer> serializerClass =
-          ((ClassResolver) typeResolver)
-              .getObjectSerializerClass(
-                  cls, sc -> dataSerializer = Serializers.newSerializer(typeResolver, cls, sc));
-      dataSerializer = Serializers.newSerializer(typeResolver, cls, serializerClass);
+      if (registerSerializer) {
+        typeResolver.setSerializer(cls, this);
+      }
+      if (typeDef == null) {
+        Class<? extends Serializer> serializerClass =
+            ((ClassResolver) typeResolver)
+                .getObjectSerializerClass(
+                    cls, sc -> dataSerializer = Serializers.newSerializer(typeResolver, cls, sc));
+        dataSerializer = Serializers.newSerializer(typeResolver, cls, serializerClass);
+      } else {
+        dataSerializer = new CompatibleSerializer<>(typeResolver, cls, typeDef);
+      }
       // No need to set object serializer to this, it will be set in class resolver later.
       // typeResolver.setSerializer(cls, this);
     }

--- a/java/fory-core/src/main/java/org/apache/fory/serializer/collection/MapSerializers.java
+++ b/java/fory-core/src/main/java/org/apache/fory/serializer/collection/MapSerializers.java
@@ -49,10 +49,12 @@ import org.apache.fory.context.CopyContext;
 import org.apache.fory.context.ReadContext;
 import org.apache.fory.context.WriteContext;
 import org.apache.fory.memory.MemoryBuffer;
+import org.apache.fory.meta.TypeDef;
 import org.apache.fory.reflect.ReflectionUtils;
 import org.apache.fory.resolver.ClassResolver;
 import org.apache.fory.resolver.TypeInfo;
 import org.apache.fory.resolver.TypeResolver;
+import org.apache.fory.serializer.CompatibleSerializer;
 import org.apache.fory.serializer.ExternalizableSerializer;
 import org.apache.fory.serializer.GraphMemoryEstimates;
 import org.apache.fory.serializer.ReplaceResolveSerializer;
@@ -510,16 +512,31 @@ public class MapSerializers {
     private Serializer<T> dataSerializer;
 
     public DefaultJavaMapSerializer(TypeResolver typeResolver, Class<T> cls) {
+      this(typeResolver, cls, null, true);
+    }
+
+    public DefaultJavaMapSerializer(TypeResolver typeResolver, Class<T> cls, TypeDef typeDef) {
+      this(typeResolver, cls, typeDef, false);
+    }
+
+    private DefaultJavaMapSerializer(
+        TypeResolver typeResolver, Class<T> cls, TypeDef typeDef, boolean registerSerializer) {
       super(typeResolver, cls, false);
       Preconditions.checkArgument(
           !config.isXlang(),
           "Fory cross-language default map serializer should use " + MapSerializer.class);
-      typeResolver.setSerializer(cls, this);
-      Class<? extends Serializer> serializerClass =
-          ((ClassResolver) typeResolver)
-              .getObjectSerializerClass(
-                  cls, sc -> dataSerializer = Serializers.newSerializer(typeResolver, cls, sc));
-      dataSerializer = Serializers.newSerializer(typeResolver, cls, serializerClass);
+      if (registerSerializer) {
+        typeResolver.setSerializer(cls, this);
+      }
+      if (typeDef == null) {
+        Class<? extends Serializer> serializerClass =
+            ((ClassResolver) typeResolver)
+                .getObjectSerializerClass(
+                    cls, sc -> dataSerializer = Serializers.newSerializer(typeResolver, cls, sc));
+        dataSerializer = Serializers.newSerializer(typeResolver, cls, serializerClass);
+      } else {
+        dataSerializer = new CompatibleSerializer<>(typeResolver, cls, typeDef);
+      }
       // No need to set object serializer to this, it will be set in class resolver later.
       // typeResolver.setSerializer(cls, this);
     }

--- a/java/fory-core/src/test/java/org/apache/fory/serializer/collection/CollectionCompatibleSerializerTest.java
+++ b/java/fory-core/src/test/java/org/apache/fory/serializer/collection/CollectionCompatibleSerializerTest.java
@@ -19,19 +19,32 @@
 
 package org.apache.fory.serializer.collection;
 
+import java.io.ByteArrayOutputStream;
+import java.lang.reflect.Field;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.AbstractMap;
 import java.util.AbstractSet;
+import java.util.Collection;
 import java.util.Iterator;
 import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
 import java.util.Map;
 import java.util.Set;
+import javax.tools.JavaCompiler;
+import javax.tools.ToolProvider;
 import org.apache.fory.Fory;
 import org.apache.fory.ForyTestBase;
+import org.apache.fory.context.MetaReadContext;
+import org.apache.fory.context.MetaWriteContext;
+import org.apache.fory.util.ClassLoaderUtils;
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
 public class CollectionCompatibleSerializerTest extends ForyTestBase {
+  private static final String PEER_PACKAGE = "org.apache.fory.serializer.collection.compat";
 
   @Test
   public void testCustomSetCompatible() {
@@ -69,8 +82,234 @@ public class CollectionCompatibleSerializerTest extends ForyTestBase {
     Assert.assertEquals(copyMap, map);
   }
 
+  @Test
+  public void testDefaultSetEvolution() throws Exception {
+    Class<?> writerPeer = compilePeer("DefaultSetPeer", defaultSetSource(true));
+    Class<?> readerPeer = compilePeer("DefaultSetPeer", defaultSetSource(false));
+    Object writerHolder = writerPeer.getMethod("newHolder").invoke(null);
+
+    Object copy = peerSerDe(writerPeer.getClassLoader(), readerPeer.getClassLoader(), writerHolder);
+    Object values = fieldValue(copy, "values");
+    Assert.assertEquals(values.getClass().getName(), PEER_PACKAGE + ".DefaultSetPeer$CustomSet");
+    Assert.assertEquals(fieldValue(values, "marker"), "set-marker");
+    Assert.assertEquals(((Collection<?>) values).size(), 2);
+    Assert.assertTrue(((Collection<?>) values).contains("a"));
+  }
+
+  @Test
+  public void testDefaultMapEvolution() throws Exception {
+    Class<?> writerPeer = compilePeer("DefaultMapPeer", defaultMapSource(true));
+    Class<?> readerPeer = compilePeer("DefaultMapPeer", defaultMapSource(false));
+    Object writerHolder = writerPeer.getMethod("newHolder").invoke(null);
+
+    Object copy = peerSerDe(writerPeer.getClassLoader(), readerPeer.getClassLoader(), writerHolder);
+    Object values = fieldValue(copy, "values");
+    Assert.assertEquals(values.getClass().getName(), PEER_PACKAGE + ".DefaultMapPeer$CustomMap");
+    Assert.assertEquals(fieldValue(values, "marker"), "map-marker");
+    Assert.assertEquals(((Map<?, ?>) values).get("k1"), "v1");
+    Assert.assertEquals(((Map<?, ?>) values).size(), 2);
+  }
+
+  @Test
+  public void testChildSetEvolution() throws Exception {
+    Class<?> writerPeer = compilePeer("ChildSetPeer", childSetSource(true));
+    Class<?> readerPeer = compilePeer("ChildSetPeer", childSetSource(false));
+    Object writerHolder = writerPeer.getMethod("newHolder").invoke(null);
+
+    Object copy = peerSerDe(writerPeer.getClassLoader(), readerPeer.getClassLoader(), writerHolder);
+    Object values = fieldValue(copy, "values");
+    Assert.assertEquals(values.getClass().getName(), PEER_PACKAGE + ".ChildSetPeer$CustomSet");
+    Assert.assertEquals(fieldValue(values, "marker"), "child-set-marker");
+    Assert.assertTrue(((Collection<?>) values).contains("a"));
+  }
+
+  @Test
+  public void testChildMapEvolution() throws Exception {
+    Class<?> writerPeer = compilePeer("ChildMapPeer", childMapSource(true));
+    Class<?> readerPeer = compilePeer("ChildMapPeer", childMapSource(false));
+    Object writerHolder = writerPeer.getMethod("newHolder").invoke(null);
+
+    Object copy = peerSerDe(writerPeer.getClassLoader(), readerPeer.getClassLoader(), writerHolder);
+    Object values = fieldValue(copy, "values");
+    Assert.assertEquals(values.getClass().getName(), PEER_PACKAGE + ".ChildMapPeer$CustomMap");
+    Assert.assertEquals(fieldValue(values, "marker"), "child-map-marker");
+    Assert.assertEquals(((Map<?, ?>) values).get("k1"), "v1");
+  }
+
+  @Test
+  public void testObjectStreamEvolution() throws Exception {
+    Class<?> writerPeer = compilePeer("StreamPeer", streamSource(true));
+    Class<?> readerPeer = compilePeer("StreamPeer", streamSource(false));
+    Object writerValue = writerPeer.getMethod("newValue").invoke(null);
+
+    Object copy = peerSerDe(writerPeer.getClassLoader(), readerPeer.getClassLoader(), writerValue);
+    Assert.assertEquals(copy.getClass().getName(), PEER_PACKAGE + ".StreamPeer");
+    Assert.assertEquals(fieldValue(copy, "marker"), "stream-marker");
+  }
+
   private static Fory compatibleFory() {
     return builder().withRefTracking(true).withCodegen(true).withCompatible(true).build();
+  }
+
+  private static Fory peerFory(ClassLoader classLoader) {
+    return builder()
+        .withClassLoader(classLoader)
+        .withRefTracking(true)
+        .withCodegen(true)
+        .withCompatible(true)
+        .withMetaShare(true)
+        .withScopedMetaShare(false)
+        .build();
+  }
+
+  private static Object peerSerDe(
+      ClassLoader writerClassLoader, ClassLoader readerClassLoader, Object value) {
+    Fory writer = peerFory(writerClassLoader);
+    Fory reader = peerFory(readerClassLoader);
+    MetaWriteContext metaWriteContext = new MetaWriteContext();
+    MetaReadContext metaReadContext = new MetaReadContext();
+    setMetaContexts(writer, metaWriteContext, new MetaReadContext());
+    byte[] bytes = writer.serialize(value);
+    setMetaContexts(reader, new MetaWriteContext(), metaReadContext);
+    return reader.deserialize(bytes);
+  }
+
+  private static Object fieldValue(Object obj, String name) throws ReflectiveOperationException {
+    Field field = obj.getClass().getField(name);
+    return field.get(obj);
+  }
+
+  private static Class<?> compilePeer(String className, String source) throws Exception {
+    Path dir = Files.createTempDirectory("fory-compatible-collection");
+    Path sourceFile = dir.resolve(PEER_PACKAGE.replace('.', '/')).resolve(className + ".java");
+    Files.createDirectories(sourceFile.getParent());
+    Files.write(sourceFile, source.getBytes(StandardCharsets.UTF_8));
+    JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
+    Assert.assertNotNull(compiler, "Tests require a JDK compiler");
+    int result =
+        compiler.run(
+            null,
+            new ByteArrayOutputStream(),
+            System.err,
+            "-classpath",
+            System.getProperty("java.class.path"),
+            "-d",
+            dir.toString(),
+            sourceFile.toString());
+    if (result != 0) {
+      throw new AssertionError("Couldn't compile code:\n" + source);
+    }
+    URL[] urls = new URL[] {dir.toUri().toURL()};
+    ClassLoader parent = CollectionCompatibleSerializerTest.class.getClassLoader();
+    return new ClassLoaderUtils.ChildFirstURLClassLoader(urls, parent)
+        .loadClass(PEER_PACKAGE + "." + className);
+  }
+
+  private static String defaultSetSource(boolean writer) {
+    return sourceHeader("DefaultSetPeer")
+        + "  public static class Holder { public Set values; }\n"
+        + "  public static class CustomSet extends AbstractSet {\n"
+        + "    public LinkedHashSet values = new LinkedHashSet();\n"
+        + "    public String marker;\n"
+        + (writer ? "    public String writerOnly;\n" : "")
+        + "    public Iterator iterator() { return values.iterator(); }\n"
+        + "    public int size() { return values.size(); }\n"
+        + "    public boolean add(Object value) { return values.add(value); }\n"
+        + "  }\n"
+        + "  public static Holder newHolder() {\n"
+        + "    Holder holder = new Holder();\n"
+        + "    CustomSet values = new CustomSet();\n"
+        + "    values.marker = \"set-marker\";\n"
+        + (writer ? "    values.writerOnly = \"writer-only\";\n" : "")
+        + "    values.add(\"a\"); values.add(\"b\"); holder.values = values; return holder;\n"
+        + "  }\n"
+        + "}\n";
+  }
+
+  private static String defaultMapSource(boolean writer) {
+    return sourceHeader("DefaultMapPeer")
+        + "  public static class Holder { public Map values; }\n"
+        + "  public static class CustomMap extends AbstractMap {\n"
+        + "    public LinkedHashMap values = new LinkedHashMap();\n"
+        + "    public String marker;\n"
+        + (writer ? "    public String writerOnly;\n" : "")
+        + "    public Object put(Object key, Object value) { return values.put(key, value); }\n"
+        + "    public Set entrySet() { return values.entrySet(); }\n"
+        + "  }\n"
+        + "  public static Holder newHolder() {\n"
+        + "    Holder holder = new Holder();\n"
+        + "    CustomMap values = new CustomMap();\n"
+        + "    values.marker = \"map-marker\";\n"
+        + (writer ? "    values.writerOnly = \"writer-only\";\n" : "")
+        + "    values.put(\"k1\", \"v1\"); values.put(\"k2\", \"v2\");\n"
+        + "    holder.values = values; return holder;\n"
+        + "  }\n"
+        + "}\n";
+  }
+
+  private static String childSetSource(boolean writer) {
+    return sourceHeader("ChildSetPeer")
+        + "  public static class Holder { public Set values; }\n"
+        + "  public static class CustomSet extends HashSet {\n"
+        + "    public String marker;\n"
+        + (writer ? "    public String writerOnly;\n" : "")
+        + "  }\n"
+        + "  public static Holder newHolder() {\n"
+        + "    Holder holder = new Holder();\n"
+        + "    CustomSet values = new CustomSet();\n"
+        + "    values.marker = \"child-set-marker\";\n"
+        + (writer ? "    values.writerOnly = \"writer-only\";\n" : "")
+        + "    values.add(\"a\"); values.add(\"b\"); holder.values = values; return holder;\n"
+        + "  }\n"
+        + "}\n";
+  }
+
+  private static String childMapSource(boolean writer) {
+    return sourceHeader("ChildMapPeer")
+        + "  public static class Holder { public Map values; }\n"
+        + "  public static class CustomMap extends HashMap {\n"
+        + "    public String marker;\n"
+        + (writer ? "    public String writerOnly;\n" : "")
+        + "  }\n"
+        + "  public static Holder newHolder() {\n"
+        + "    Holder holder = new Holder();\n"
+        + "    CustomMap values = new CustomMap();\n"
+        + "    values.marker = \"child-map-marker\";\n"
+        + (writer ? "    values.writerOnly = \"writer-only\";\n" : "")
+        + "    values.put(\"k1\", \"v1\"); values.put(\"k2\", \"v2\");\n"
+        + "    holder.values = values; return holder;\n"
+        + "  }\n"
+        + "}\n";
+  }
+
+  private static String streamSource(boolean writer) {
+    return sourceHeader("StreamPeer")
+        + "  public String marker;\n"
+        + (writer ? "  public String writerOnly;\n" : "")
+        + "  private void writeObject(ObjectOutputStream out) throws IOException {\n"
+        + "    out.defaultWriteObject();\n"
+        + "  }\n"
+        + "  private void readObject(ObjectInputStream in)\n"
+        + "      throws IOException, ClassNotFoundException {\n"
+        + "    in.defaultReadObject();\n"
+        + "  }\n"
+        + "  public static StreamPeer newValue() {\n"
+        + "    StreamPeer value = new StreamPeer(); value.marker = \"stream-marker\";\n"
+        + (writer ? "    value.writerOnly = \"writer-only\";\n" : "")
+        + "    return value;\n"
+        + "  }\n"
+        + "}\n";
+  }
+
+  private static String sourceHeader(String className) {
+    return "package "
+        + PEER_PACKAGE
+        + ";\n"
+        + "import java.io.*;\n"
+        + "import java.util.*;\n"
+        + "public class "
+        + className
+        + " {\n";
   }
 
   public static final class SetHolder {

--- a/java/fory-core/src/test/java/org/apache/fory/serializer/collection/CollectionCompatibleSerializerTest.java
+++ b/java/fory-core/src/test/java/org/apache/fory/serializer/collection/CollectionCompatibleSerializerTest.java
@@ -1,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.fory.serializer.collection;
+
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+import org.apache.fory.Fory;
+import org.apache.fory.ForyTestBase;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class CollectionCompatibleSerializerTest extends ForyTestBase {
+
+  @Test
+  public void testCustomSetCompatible() {
+    Fory writer = compatibleFory();
+    Fory reader = compatibleFory();
+    CustomSet set = new CustomSet("set-marker");
+    set.add("a");
+    set.add("b");
+
+    SetHolder holder = new SetHolder();
+    holder.values = set;
+
+    SetHolder copy = serDe(writer, reader, holder);
+    Assert.assertEquals(copy.values.getClass(), CustomSet.class);
+    CustomSet copySet = (CustomSet) copy.values;
+    Assert.assertEquals(copySet.getMarker(), "set-marker");
+    Assert.assertEquals(copySet, set);
+  }
+
+  @Test
+  public void testCustomMapCompatible() {
+    Fory writer = compatibleFory();
+    Fory reader = compatibleFory();
+    CustomMap map = new CustomMap("map-marker");
+    map.put("k1", "v1");
+    map.put("k2", "v2");
+
+    MapHolder holder = new MapHolder();
+    holder.values = map;
+
+    MapHolder copy = serDe(writer, reader, holder);
+    Assert.assertEquals(copy.values.getClass(), CustomMap.class);
+    CustomMap copyMap = (CustomMap) copy.values;
+    Assert.assertEquals(copyMap.getMarker(), "map-marker");
+    Assert.assertEquals(copyMap, map);
+  }
+
+  private static Fory compatibleFory() {
+    return builder().withRefTracking(true).withCodegen(true).withCompatible(true).build();
+  }
+
+  public static final class SetHolder {
+    public Set<String> values;
+  }
+
+  public static final class MapHolder {
+    public Map<String, String> values;
+  }
+
+  public static final class CustomSet extends AbstractSet<String> {
+    private LinkedHashSet<String> values = new LinkedHashSet<>();
+    private String marker;
+
+    public CustomSet() {}
+
+    private CustomSet(String marker) {
+      this.marker = marker;
+    }
+
+    public String getMarker() {
+      return marker;
+    }
+
+    @Override
+    public boolean add(String value) {
+      return values.add(value);
+    }
+
+    @Override
+    public Iterator<String> iterator() {
+      return values.iterator();
+    }
+
+    @Override
+    public int size() {
+      return values.size();
+    }
+  }
+
+  public static final class CustomMap extends AbstractMap<String, String> {
+    private LinkedHashMap<String, String> values = new LinkedHashMap<>();
+    private String marker;
+
+    public CustomMap() {}
+
+    private CustomMap(String marker) {
+      this.marker = marker;
+    }
+
+    public String getMarker() {
+      return marker;
+    }
+
+    @Override
+    public String put(String key, String value) {
+      return values.put(key, value);
+    }
+
+    @Override
+    public Set<Entry<String, String>> entrySet() {
+      return values.entrySet();
+    }
+  }
+}

